src/recordlinker/linking/matchers.py

```diff
@@ -18,10 +19,54 @@
 
 SIMILARITY_MEASURES = typing.Literal["JaroWinkler", "Levenshtein", "DamerauLevenshtein"]
 
-# A Callable type for comparing a feature on two records
-FEATURE_COMPARE_FUNC = typing.Callable[[PIIRecord, Patient, Feature, typing.Any], float]
-# A Callable type for evaluating whether a set of feature comparisons constitutes a match
-MATCH_RULE_FUNC = typing.Callable[[list[float], typing.Any], bool]
+
+class RuleFunc(enum.Enum):
+    """
+    Enum for the different types of match rules that can be used for patient
+    matching. This is the universe of all possible match rules that a user can
+    choose from when configuring their algorithm.  When data is loaded into the
+    MPI, all possible RuleFuncs will be created for the defined match rules.
+    However, only a subset will be used in matching, based on the configuration of
+    the algorithm.
+    """
+
+    PERFECT_MATCH = "func:recordlinker.linking.matchers.eval_perfect_match"
+    LOG_ODDS_CUTOFF = "func:recordlinker.linking.matchers.eval_log_odds_cutoff"
+
+
+class FeatureFunc(enum.Enum):
+    """
+    Enum for the different types of feature comparison functions that can be used
+    for patient matching. This is the universe of all possible feature comparison
+    functions that a user can choose from when configuring their algorithm.  When
+    data is loaded into the MPI, all possible FeatureFuncs will be created for the
+    defined feature comparison functions. However, only a subset will be used in
+    matching, based on the configuration of the algorithm.
+    """
+
+    MATCH_ANY = "func:recordlinker.linking.matchers.feature_match_any"
+    MATCH_EXACT = "func:recordlinker.linking.matchers.feature_match_exact"
+    MATCH_FUZZY_STRING = "func:recordlinker.linking.matchers.feature_match_fuzzy_string"
+    MATCH_LOG_ODDS_FUZZY_COMPARE = (
+        "func:recordlinker.linking.matchers.feature_match_log_odds_fuzzy_compare"
+    )
+
+
+class AvailableKwarg(enum.Enum):
+    """
+    Enum for the different types of keyword arguments that can be used in the
+    AlgorithmPass schema. This is the universe of all possible keyword arguments
+    that a user can choose from when configuring their algorithm.  When data is
+    loaded into the MPI, all possible AvailableKwargs will be created for the
+    defined keyword arguments. However, only a subset will be used in matching,
+    based on the configuration of the algorithm.
+    """
+
+    SIMILARITY_MEASURE = "similarity_measure"
+    THRESHOLD = "threshold"
+    THRESHOLDS = "thresholds"
```

Comment:
What is the difference between these kwargs? If they are thresholds for different parameters, maybe we should distinguish them by name here. I can also include this in the renaming ticket #109

Reply:
This is a [holdover from the original implementation](https://github.com/CDCgov/phdi/blob/main/containers/record-linkage/app/linkage/link.py#L1021-L1025).  From what I can tell `threshold` points to a single float value that is to be used for all features, `thresholds` points to a mapping of features-> floats and used for feature specific thresholds.

In general, I'd like to rethink everything about kwargs as I think it will be difficult for users to remember when they have to specify them, and what values they need as they use different functions.  There is also the chance that some users might want different values in a pass depending on which evaluator step they are in.